File: 2008math4.rb
#!/usr/bin/ruby
# coding: utf-8
#
# 出題元
# 麻布中学校 2008年 入試問題 算数 問4 解法

class ExprSet
  attr_reader :exprSet

  def initialize(minNum, maxNum)
    @exprSet = ((minNum+1)..maxNum).inject([minNum]) do |xs, i|
      xs.product(["+", "*"], [i])
    end.map(&:join)
  end

  def printSums
    puts @exprSet.map{ |expr| "#{eval(expr)} = #{expr}"}.sort_by(&:to_i)
  end

  def printMatchedSums(other)
    puts @exprSet.product(other.exprSet).map { |x| x.join(" == ") }.select { |x| eval x }
  end

  # より詳しく報告する
  def printMatchedSumsAndValue(other)
    @exprSet.product(other.exprSet).map do |l,r|
      [l, [l, r].join(" == ")]
    end.select { |l,expr| eval expr }.each do |l,expr|
      puts (eval l).to_s + " : " + expr
    end
  end
end

if (ARGV.size >= 4)
  v = ARGV.map(&:to_i)
  ExprSet.new(v[0],v[1]).printMatchedSumsAndValue(ExprSet.new(v[2],v[3]))
else
  # 問4-1
  ExprSet.new(1,4).printSums
  # 問4-2
  ExprSet.new(1,5).printMatchedSums(ExprSet.new(2,6))
end

exit(0)

File: 2008math4.pl
#!/usr/bin/perl
# coding: utf-8
#
# 出題元
# 麻布中学校 2008年 入試問題 算数 問4 解法

sub expressions {
    my ($minNum, $maxNum) = @_;
    my @ls;
    my $num = "" . $minNum;

    if ($minNum == $maxNum) {
        push(@ls, $num);
    } else {
        foreach my $subExpr (expressions($minNum + 1, $maxNum)) {
           push(@ls, $num . "+" . $subExpr);
           push(@ls, $num . "*" . $subExpr);
        }
    }

    @ls;
}

sub printSums {
    my ($minNum, $maxNum) = @_;
    my @results;
    foreach my $expr (expressions($minNum, $maxNum)) {
        my $result = eval($expr);
        push(@results, ($result . " = " . $expr));
    }

    foreach my $result (sort { $a <=> $b } @results) {
        print $result . "\n";
    }
}

sub printMatchedSums {
    my ($minLeft, $maxLeft, $minRight, $maxRight) = @_;
    foreach my $leftExpr (expressions($minLeft, $maxLeft)) {
        foreach my $rightExpr (expressions($minRight, $maxRight)) {
            if (eval($leftExpr) == eval($rightExpr)) {
                print $leftExpr . " == " . $rightExpr . "\n";
            }
        }
    }
}

printSums(1, 4);
printMatchedSums(1, 5, 2, 6);
0;
